src/runtime/virtcontainers/kata_agent.go

```diff
@@ -1609,29 +1606,33 @@ func (k *kataAgent) handleBlockVolumes(c *Container) ([]*grpc.Storage, error) {
 		// device is detached with detachDevices() for a container.
 		c.devices = append(c.devices, ContainerDevice{ID: id, ContainerPath: m.Destination})
 
-		var vol *grpc.Storage
-
-		device := c.sandbox.devManager.GetDeviceByID(id)
-		if device == nil {
-			k.Logger().WithField("device", id).Error("failed to find device by id")
-			return nil, fmt.Errorf("Failed to find device by id (id=%s)", id)
+		// Create Storage structure
+		vol, err := k.createBlkStorageObject(c, m)
+		if vol == nil || err != nil {
+			return nil, err
 		}
 
-		var err error
-		switch device.DeviceType() {
-		case config.DeviceBlock:
-			vol, err = k.handleDeviceBlockVolume(c, m, device)
-		case config.VhostUserBlk:
-			vol, err = k.handleVhostUserBlkVolume(c, m, device)
-		default:
-			k.Logger().Error("Unknown device type")
-			continue
-		}
+		// The device will be mounted at a unique location within the VM. Mounting
+		// to the container specific location is handled within the OCI spec. Let's ensure that
+		// the storage mount point is unique, and that this is utilized as the source in the OCI
+		// spec.
+		filename := fmt.Sprintf("%s-%s", uuid.Generate().String(), filepath.Base(vol.MountPoint))
```

Comment:
If two containers within a pod wish to mount the same block device based volume, does this work out? Should the generation of the unique filename (where the file system will be mounted to in the guest) be done during sandbox creation rather than container creation?

Reply:
@ddebroy Good catch! But I think it might be safe to mount the same filesystem onto the device twice. I tried to mount a loopback device to two different mount points, it appears both mount points have the same content. 
@bergwolf  @egernst wdyt?

Reply:
Deep, I think your premise is good here, thought we don't know about the volumes at sandbox creation time, only at sandbox.

IMO, the sandbox should be in charge of adding the volume, and at container creation, it should call into sandbox to say "please make this volume available to the guest."

The sandbox then should have a refcount for each volume _source_. If this volume _source_ is new, we should be handling appropriately (either hot plug a new blk device, or bind mount into the virtiofs shared directory), along with directions on where to make available in guest (either by mounting, or based on path from virtiofs shared directory).  If the volume already exists, increment refcount. In either case, finish by returning the in-VM location of the volume to create Container so it can then update the in-guest OCI spec to reflect the in-guest mount source.

Reply:
@egernst @ddebroy @bergwolf and I had an offline discussion on this. The conclusion is, the CSI driver should be able to indicate appropriate mount permission (and the permission for device passing to hypervisor). The volume should be refcounted and the filesystem should be only mounted once inside the guest. The changes have been made in the latest commit of this PR.